feat : Restaurant 엔티티 구현

hyerijang opened the pull request:
## 🚀 풀 리퀘스트 요약

@seoyoon047 님 API 기능 구현하실 때 필요하실 것 같아 우선 구현한 부분 올립니다. 

`Restaurant 엔티티`
- raw 데이터의 모든 필드를 저장합니다 - **상의 후 필요한 필드만 저장할 지 고려** 
- Location, Address 와 1대 1관계로 맵핑됩니다.
- N+1 문제 예방을 위해, @xToOne 연관관계인 엔티티는 서로 Lazy로딩하도록 설정하였습니다.
- 양방향 연관관계에서 엔티티 -> json 변환시 발생할 수 있는 무한루프를 예방하기 위해 @JsonIgnore을 설정 하였습니다.

## 📋 변경 사항

- [x] 새로운 기능 추가
- [x] 테스트 추가, 테스트 리팩토링


## 📸 스크린샷

![image](https://github.com/wanted-quantum-jump/FoodieFinder/assets/46921979/41fc60bb-c57e-4324-b6da-890eb33d7da6)

## 📌 체크리스트
- [x] 모든 테스트를 통과합니다.

## 📎 관련 이슈

#2 

## 🙌 리뷰 및 피드백

hyerijang committed:
feat : Restaurant 엔티티 구현

- 모든 raw 데이터를 저장합니다.
- Location, Address 와 1대 1관계로 맵핑됩니다.
- N+1 문제 예방을 위해, @xToOne 연관관계인 엔티티는 서로 Lazy로딩하도록 설정하였습니다.
- 양방향 연관관계에서 엔티티 -> json 변환시 발생할 수 있는 무한루프를 예방하기 위해 @JsonIgnore을 설정 하였습니다.

#2

hyerijang commented:
@seoyoon047 
- `Location`, `Address`에 속한 필드들은 서비스 특성 상(`지리기반` 맛집 추천) 조회할 일이 잦을 것 같아 성능을 위해 테이블을 분리했습니다. 

- `Restaurant - Location`, `Restaurant - Address`가 양방향 연관관계이긴 하지만 
  - object -> json 변환시 발생하는 직렬화 문제는  @ JsonIgnore로 예방해 뒀고, 
  - 애초에 DTO를 써서 엔티티 자체를 노출(respone 혹은 json화 ) 시키지만 않으면 되는 문제라 
- 괜찮을거라고 생각합니다!

이대로 구현해도 괜찮을까요? 테이블이 너무 많아져서 복잡하다거나, 사용하기 불편하다 하시면 그냥 다시 한 테이블로 합치겠습니다. 합치는건 금방하니 편하게 말해주세요. 😀

hyerijang commented:
지금 남성종사자수(명) 처럼 다소 필요없는 필드가 저장되고 있는데, `어떤 필드를 빼면 좋을지` 의견 바랍니다. 
(raw데이터는 Rawdata 테이블에 따로 저장하고 Restaurant에서는 실제 API에서 이용될 데이터만 저장하려고 합니다)

O0oO0Oo commented:
> 지금 남성종사자수(명) 처럼 다소 필요없는 필드가 저장되고 있는데, 어떤 필드를 빼면 좋을지 의견 바랍니다. (raw데이터는 Rawdata 테이블에 따로 저장하고 Restaurant에서는 실제 API에서 이용될 데이터만 저장하려고 합니다)

- 시군명(SIGUN_NM)
- 사업장명(BIZPLC_NM)
- 영업상태명(BSN_STATE_NM) 
- 주소(우편주소와 도로명/지명주소 는 둘 다?)
- 위도, 경도

전 이정도로 보이네요

hyerijang commented:
> > 지금 남성종사자수(명) 처럼 다소 필요없는 필드가 저장되고 있는데, 어떤 필드를 빼면 좋을지 의견 바랍니다. (raw데이터는 Rawdata 테이블에 따로 저장하고 Restaurant에서는 실제 API에서 이용될 데이터만 저장하려고 합니다)
> 
> * 시군명(SIGUN_NM)
> * 사업장명(BIZPLC_NM)
> * 영업상태명(BSN_STATE_NM)
> * 주소(우편주소와 도로명/지명주소 는 둘 다?)
> * 위도, 경도
> 
> 전 이정도로 보이네요

네 좋습니다! 그럼 테이블 크기가 중분히 작아져서 주소랑 위치 테이블을 굳이 분리할 필요가 없겠네요. 그냥 Restaurant 하나로 통합시키겠습니다.

hyerijang commented:
![image](https://github.com/wanted-quantum-jump/FoodieFinder/assets/46921979/a4574976-3ce6-42b2-8209-3c25d52522b5)
**Restaurant 테이블**
- 시군명(SIGUN_NM)
- 사업장명(BIZPLC_NM)
- 영업상태명(BSN_STATE_NM)
- 사업장명(BIZPLC_NM)
- 영업상태명(BSN_STATE_NM)
- 위생업태명(SANITTN_BIZCOND_NM) : **카페, 김밥(도시락) 등 업종을 나타내는 필드라 추가하였습니다.**
- 소재지도로명주소(REFINE_ROADNM_ADDR)
- 소재지지번주소(REFINE_LOTNO_ADDR)
- 소재지우편번호 (REFINE_ZIP_CD)
- WGS84위도(REFINE_WGS84_LAT)
- WGS84경도(REFINE_WGS84_LOGT)
- 해당 맛집의 평균 평점


우선 이렇게 확정 짓겠습니다. @seoyoon047 님이 API 쓰실때 평점 엔티티랑 연관관계 맺어서 알아서 쓰시면 될 것 같아요!

hyerijang committed:
refactor : Restaurant 엔티티 수정

- raw 데이터가 아닌 API에서 실제 사용할 필드만 저장합니다.
- Location, Address 테이블은 삭제하고 Restaurant로 통합하였습니다.

#2

hyerijang committed:
feat : 리포지토리 단위 테스트를 위한 커스텀 어노테이션 생성

hyerijang committed:
test : Restaurant 제약조건 단위 테스트

- 제약조건을 만족하는 엔티티만 저장되는지 보기위해 RestaurantRepositoryTest 단위 테스트를 작성했습니다.

hyerijang committed:
refactor : Restaurant 빌더 생성자에서 averageRating 제외

- 엔티티 생성시에는 평균 평점이 존재하지 않으므로 빌더에서 제외하였습니다

hyerijang committed:
style : BaseTimeEntity 주석 추가

클래스 설명을 추가했습니다.

hyerijang committed:
feat : RawRestaurant 추가

raw 데이터 저장을 위한 엔티티입니다,

hyerijang committed:
test : RawRestaurant 저장 및 유니크 조건 단위 테스트

hyerijang committed:
style : 주석 추가 및 DisplayName 내용 변경

cjw9506 commented:
초기 설정에 "Restaurant - Location, Restaurant - Address가 양방향 연관관계이긴 하지만" 이라고 작성해주셨는데,
이 부분은 단방향으로도 충분하지 않을까요?
합치는 방향으로 하시긴했지만, 초기에 단방향말고 양방향으로 설정하신 이유가 궁금합니다.

hyerijang commented:
> 초기 설정에 "Restaurant - Location, Restaurant - Address가 양방향 연관관계이긴 하지만" 이라고 작성해주셨는데, 이 부분은 단방향으로도 충분하지 않을까요? 합치는 방향으로 하시긴했지만, 초기에 단방향말고 양방향으로 설정하신 이유가 궁금합니다.

Location 하나로만 예시를 들겠습니다

**Restaurant → Location 가 필요한 이유**

- Restaurant  `상세 정보 조회` 시 해당`음식점`의 `위치`를 가져와야함

**Location → Restaurant 이 필요한 이유**

- `맛집 목록(API)` : 유저의 위치를 기준으로 `근방`에 있는 `음식점`을 조회
    - 모든 Location 중 해당 조건에 일치하는  Location 를 찾아서
        
        List<Location> locationList = locationRepositor.(이하생략)
        
    - 해당 Location과 연관관계인 Restaurant 와 함께 반환.
        - **이때 Location → Restaurant 필요**

이런 이유로 양방향 연관관계로 설정해 놓았습니다.

`무한 참조` (Location → Restaurant →  Location  같은 상황)의 경우 사용자에게 Restaurant 나 Location 엔티티를 직접 Response하지 않으면 발생하지 않기때문에, (응답시, Jackson 라이브러리가 엔티티를  json으로 변환하는 과정에서 연관관계인 엔티티를 계속 타고 내려가면서 발생)

`DTO`로 변환해서 응답하는 저희 팀에서는 발생하지 않을 문제라고 생각했습니다.  그래도 혹시 엔티티가 노출되어서 문제가 될까봐 @ JsonIgnore로 무한참조는 막아두었습니다.

seoyoon047 commented:
깔끔하게 잘 설계하셨습니다 👍 Merge하겠습니다
